### DISCUSSION TITLE
2024/06/03/install-mysql-via-docker/

### DISCUSSION THREAD

**Discussion by giscus** [cyclinbox/cyclinbox.github.io#23] [General]
# 2024/06/03/install-mysql-via-docker/

由于工作需要，我要在一个docker环境中安装一个mysql，于是对docker和mysql做了一点探究，笔记如下。

https://wz.anoms.top/2024/06/03/install-mysql-via-docker/

**Comment by Hazelandinstant** [comment]
我之前就想用docker但一直没搞清楚
所以我看下来觉得docker的主要适用场所是：当你没有root权限而又要使用服务器中需要root权限简单修改调用的程序时，可以通过docker来做一个镜像，设置相应参数做一个容器，来解决这个问题，是吗？

但我还是不太理解：如果不能在root权限下使用这个软件，那么是否可以放在自己目录下呢？是因为这样影响软件运行吗？比如必须调用一些root权限才能运行的部分？

可是你前面提出的docker是一个虚拟化、沙箱机制的容器，所以是可以通过docker有限访问一些root里的权限吗（通过端口）？
先提前感谢blog主的解答 :P

**Comment by cyclinbox** [comment]
关于你的第一个问题，这确实是docker使用场景之一，但docker的使用场景不止这一个。由于docker内的环境和宿主机互相隔离，在这里运行程序更加安全，因此常用于开发和测试。此外，docker支持构建和打包镜像（把容器内的各种环境变量、依赖库和文件全打包在一起），因此也很适合软件分发（别人用这个软件的时候只需要docker pull下来就能运行，几乎不会出现缺乏依赖库或者版本冲突的问题）。

对于第二个问题，“如果不能在root权限下使用这个软件，那么是否可以放在自己目录下呢？”，这得分情况讨论。有些软件需要root权限才能运行，因为它需要一些系统目录下的资源（例如`/usr/bin`、`/usr/lib`），这种情况除了docker技术以外也可以通过chroot技术实现，后者通过重定向进程及其子进程的root目录到文件系统上的新位置，从而分离每个进程的文件访问权限，使得该进程无法接触到外部的文件（因此可以像你说的那样，把软件放在自己的目录下）。然而，另一些软件（例如本文里的mysql-server），其需要借助systemctl启动系统级别的服务进程，这种高权限的进程需要root身份的用户才能启动，因此只能使用docker或者虚拟机一类的技术。

对于最后一个问题，“通过docker有限访问一些root里的权限”有一部分是宿主机内docker本身的权限（docker程序的权限比root低一点，但是比普通用户高），另外在容器内部root权限是完整的。也因此，docker可以使用端口实现容器内外的通信。
